### DISCUSSION TITLE
[Chapter 09] 예외를 catch 하고 무시하는 정도는 어디까지인가?

### DISCUSSION THREAD

**Discussion by corock** [cowaine/is-my-code-that-strange#80] [Chapter 09-10 (악마, 이름 설계)] [answered]
이커머스 프로젝트에서 다음과 같은 요구사항이 있다고 가정해 보겠습니다.

- 특정 상품을 주문하기 위해 주문서 진입 시, 적립 예정 포인트(`EstimatedRewardPoints`)가 얼마인지 고객에게 제공한다.
- 적립 예정 포인트는 포인트 서버를 통해 조회할 수 있으며, 주문 서버는 회원의 통합 회원 번호(`IntegratedMemberId`)를 요청 값으로 전달한다.

위와 같은 내용을 토대로 간단하게 예시를 작성해 보면 다음과 같습니다.

```java
@Slf4j
@Service
@RequiredArgsConstructor
public class PointService {

    private final PointAdaptor pointAdaptor;

    public BigDecimal retrieveEstimatedRewardPoints(Member member) {
        return pointAdaptor.getRewardBeautyPoint(member.getIntegratedMemberId());
    }

}
```

그런데 어느 날 적립 예정 포인트를 조회하기 위한 포인트 서버가 뻗는 바람에 주문서 진입 시 오류가 발생하여 흰 화면으로 나오게 되었습니다.
주문이 아예 되지 않는 것은 매출에도 타격이 있으므로 더 이상의 장애 전파를 막기 위해 아래처럼 긴급하게 수정 후 배포하고 모니터링하기로 했습니다.

```java
@Slf4j
@Service
@RequiredArgsConstructor
public class PointService {

    private final PointAdaptor pointAdaptor;

    public BigDecimal retrieveEstimatedRewardPoints(Member member) {
        try {
            return pointAdaptor.getRewardBeautyPoint(member.getIntegratedMemberId());
        } catch (Exception e) {
            // 이 경우는 예외를 무시한 것인가?
            log.error("포인트 API 호출 실패", e);
        }

        return BigDecimal.ZERO;
    }

}
```

위 예시는 예외가 발생하면 오류 로그를 남기고 0 포인트로 반환하도록 처리한 코드입니다.
`catch` 블록에서는 어떤 예외를 던지지 않고 단순히 로그만 남기고 있습니다. 여기서 질문입니다.

1. 이 방법은 적절한 대처일까요? 적절하다면 그 이유를 간단하게 남겨주세요.
2. 만약 아니라면 다른 좋은 방법이 있을까요?

**Comment by snoopingh** [comment]
1. 굉장히 적절하지 않습니다.
2. 주문서 화면으로는 그냥 넘어가게 놔두고 적립 예정 포인트 조회에 대한 부분은 비동기로 처리하면 될 것 같습니다.

**Reply by corock** [reply]
비동기 처리는 상황에 따라 고려해 보면 좋은 아이디어인 것 같네요~

**Comment by firewoody237** [comment, marked as answer]
좋지 않다.
- 위 로그에 따르면 사후적립조차 불가능, 비즈니스적으로 타 업무에도 영향을 미침
- 적립 포인트 0만 가지고는 화면단에서 고객에게 문제때문제 포인트가 0인지, 그냥 0인지 안내 불가 -> 컴플레인 유발

MQ나 DocumentDB 같은것으로 우선 쌓고, 배치처리
혹은 로그를 잘남겨서 누군가 추후 사후적립작업 희생되기

**Reply by corock** [reply]
작년에 직접 겪었던 누군가 희생된 사례로...

1. 개발자 L 은 주문 시 예상 적립 포인트 계산 로직을 수정하면서 특정 조건을 만족한 회원은 과하게 포인트가 적립되는 사이드 이펙트가 발생했다.
2. 이슈 대상 목록을 추출해서 유관 부서로부터 포인트 적립 취소 및 적립 API 가이드를 받고, 한 땀 한 땀 포인트 취소 처리 및 재적립 처리를 진행했다.
3. L 은 이슈 분석, 유관 부서와 해결책 논의 및 데이터 보정 작업 등을 프로페셔널하게 마친 후 막차를 타고 퇴근하고 다음 날 여느 때와 같이 출근했다.

**Comment by Irisation23** [comment]
- [분산 서비스 환경에 대한 Circuit Breaker 적용](https://engineering.linecorp.com/ko/blog/circuit-breakers-for-distributed-services)
